### PR TITLE
chore: add trace logs for each constituent API call in consolidated API 

### LINKED ISSUE #29836: [Task]: Track performance of each individual response in consolidated API flow
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

https://theappsmith.slack.com/archives/C024GUDM0LT/p1702625681545779

### PR DESCRIPTION
**Pull request by sumitsum**
## Description
- Add OTLP trace log for each piece of data that the consolidated API fetches to measure the root cause of any latency regarding the consolidated API. 

#### PR fixes following issue(s)
Fixes #29836 

#### Type of change
- Chore (housekeeping or task changes that don't impact user perception)

## Testing
- Manual testing in progress
#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

## Summary by CodeRabbit

- **New Features**
  - Introduced telemetry enhancements for better tracking of user interactions and API calls.
  
- **Refactor**
  - Improved error handling and tracing logic within the application.

- **Documentation**
  - Updated internal documentation to reflect new telemetry constants and methods.

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java [hunk 1]
```diff
@@ -150,27 +176,43 @@ public Mono<ConsolidatedAPIResponseDTO> getConsolidatedInfoForPageLoad(
         ConsolidatedAPIResponseDTO consolidatedAPIResponseDTO = new ConsolidatedAPIResponseDTO();
 
         /* Get user profile data */
+        AtomicReference<Span> userProfileSpan = new AtomicReference<>();
```
Comment (rajatagrawal) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java:179]: Would this allow to correlate the constituent spans to the overall consolidated api response span ?

Can you please suggest a way how to track the overall response time of the consolidated api and the breakdown of individual components in the current approach ? Thanks!

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java [hunk 2]
```diff
@@ -206,44 +252,73 @@ public Mono<ConsolidatedAPIResponseDTO> getConsolidatedInfoForPageLoad(
...
+                            ApplicationMode.PUBLISHED.equals(mode)
+                                    ? APPLICATION_PAGES_VIEW_SPAN
+                                    : APPLICATION_PAGES_EDIT_SPAN,
+                            null));
+                })
+                .doFinally(signalType -> this.otlpTelemetry.endOtlpSpanSafely(applicationPagesSpan.get()))
                 .cache();
 
         /* Get current theme */
+        AtomicReference<Span> currentThemeSpan = new AtomicReference<>();
```
Comment (rajatagrawal) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java:287]: Can you please share insights into the requirements for using Atomic reference ? Is the same span object intended by used by multiple threads ?

If not, you may want to consider that using atomic references may affect the performance and response time of the api.

**Review by rajatagrawal — CHANGES_REQUESTED** [review]
Left review comments.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java [hunk 3]
```diff
@@ -150,27 +176,43 @@ public Mono<ConsolidatedAPIResponseDTO> getConsolidatedInfoForPageLoad(
...
+        AtomicReference<Span> tenantSpan = new AtomicReference<>();
         Mono<ResponseDTO<Tenant>> tenantResponseDTOMono = tenantService
                 .getTenantConfiguration()
                 .map(this::getSuccessResponse)
-                .onErrorResume(error -> getErrorResponseMono(error, Tenant.class));
+                .onErrorResume(error -> getErrorResponseMono(error, Tenant.class))
+                .doOnSubscribe(subscription -> {
+                    tenantSpan.set(this.otlpTelemetry.startOTLPSpan(TENANT_SPAN, null));
+                })
+                .doFinally(signalType -> this.otlpTelemetry.endOtlpSpanSafely(tenantSpan.get()));
```
[the remark this answers is not in the document]
Comment (sumitsum) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java:212]: Hi @rajatagrawal I don't think there is any other way around this. To use any variable inside a `Mono` flow it has to be effectlively final (or local variable) i.e. any non non final object cannot be used unless wrapped in `AtomicReference`. If any variable is declared as final then it cannot be assigned later i.e. assignement would have to happen outside the Mono chain which would be a weaker estimate for time bound.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java (hunk 2 above)
Comment (sumitsum) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java:287]: Hi @rajatagrawal it is a requirement imposed by usage of `lambda` expression syntax ([ref](https://www.baeldung.com/java-lambda-effectively-final-local-variables)). Encapsulating the variable within atomic reference is one way to overcome it.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java (hunk 1 above)
Comment (sumitsum) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java:179]: Thanks for the suggestion. I have added another span to track the overall API as well.

Comment (rajatagrawal) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java:179]: That wouldn't still correlate the spans for the overall api and the individual apis to each other. The requirement is that the span should have the overall time and the constitutent sub spans of it. At the moment, there is no way to tell which sub api span belongs to which overall span.

Subspans can help resolve this problem by creating a parent/child relationship between the spans. Please consider. [Reference](https://opentelemetry.io/docs/concepts/signals/traces/#context-propagation).

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java (hunk 2 above)
Comment (rajatagrawal) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ConsolidatedAPIServiceImpl.java:287]: Thanks Sumit.

It seems like that AtomicReference is used in scenarios when multiple threads want to access the same piece of code. Atomic Reference and synchronized keyword would be two ways to fix it.

Would there be a scenario here where there would be multiple threads trying to access the code here ?

As per the link shared by you, the problem identified is that the local variable may not exist in scope when the lambda function is executed. Not sure if AtomicReference would help in any way with that.

As per my understanding, making a local variable atomic makes the code really slow and should be used when it is absolutely needed.

**Comment by sumitsum** [comment]
Closing this because this PR has became too old causing many merge conflicts. It is easier to raise new one.

**Pull request closed without merging**